File: 백준/Silver/1389. 케빈 베이컨의 6단계 법칙/케빈 베이컨의 6단계 법칙.py
from collections import deque


def bfs(start):
    dist = [-1] * (N + 1)
    queue = deque([start])
    dist[start] = 0

    while queue:
        v = queue.popleft()

        for nx in graph[v]:
            if dist[nx] == -1:
                dist[nx] = dist[v] + 1
                queue.append(nx)

    return dist


N, M = map(int, input().split())

graph = [[] for _ in range(N + 1)]

for _ in range(M):
    A, B = map(int, input().split())
    graph[A].append(B)
    graph[B].append(A)

best_sum = float("inf")
best_node = 1

for i in range(1, N + 1):
    dist = bfs(i)
    total = sum(dist[1:])  # 연결 그래프라면 -1 없음
    if total < best_sum:
        best_sum = total
        best_node = i

print(best_node)

File: 백준/Silver/1389. 케빈 베이컨의 6단계 법칙/케빈 베이컨의 6단계 법칙.java
import java.io.*;
import java.util.*;

public class Main {

    private static int[] bfs(int start, List<List<Integer>> graph, int N) {
        int[] dist = new int[N + 1];
        Arrays.fill(dist, -1);

        ArrayDeque<Integer> queue = new ArrayDeque<>();
        dist[start] = 0;
        queue.offer(start);

        while (!queue.isEmpty()) {
            int v = queue.poll();

            for (int nx : graph.get(v))
                if (dist[nx] == -1) {
                    dist[nx] = dist[v] + 1;
                    queue.offer(nx);
                }
        }

        return dist;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    
        StringTokenizer st = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());

        List<List<Integer>> graph = new ArrayList<>(N + 1);
        for (int i = 0; i <= N; i++) {
            graph.add(new ArrayList<>());
        }

        for (int i = 0; i < M; i++) {
            st = new StringTokenizer(br.readLine());
            int A = Integer.parseInt(st.nextToken());
            int B = Integer.parseInt(st.nextToken());
            graph.get(A).add(B);
            graph.get(B).add(A);
        }

        int bestSum = Integer.MAX_VALUE;
        int bestNode = 1;

        // 각 노드에서 BFS -> 거리 합 계산
        for (int i = 1; i <= N; i++) {
            int[] dist = bfs(i, graph, N);

            int total = 0;
            for (int v = 1; v <= N; v++) {
                total += dist[v];
            }

            if (total < bestSum) {
                bestSum = total;
                bestNode = i;
            }
        }

        bw.write(bestNode + "\n");
        bw.flush();
    }
}   
